### PR TITLE
Add native unit tests to iOS and macOS templates

### PR DESCRIPTION
**Pull request by stuartmorgan-g**
Adds an example native unit test to the templates for iOS and macOS, matching the format we use for our 1P plugin example app unit tests. Once these have been added for all platforms+languages, they will be documented on a new plugin development page to explain their use.

For plugin example apps, the test shows an example of an actual plugin test. For standard apps, the test is an empty placeholder (similar to what Xcode would create when adding a unit test target). Although most people aren't likely to use the unit test targets in apps currently, having it be unconditional makes it much easier to maintain the project.

Also updates the Swift plugin example itself to show handling methods by name, matching the macOS Swift plugin example, since it makes it provides a better foundation for real-world plugin development (where there is very often more than one method name).

Part of https://github.com/flutter/flutter/issues/82458

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl:18]: How do you feel about this much conditionalizing in the project file? I don't love it, but I couldn't think of a better option...

File: packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl:428]: How much of this section should we prune? This is all just the defaults from Xcode (other than changing deployment back to 11).

File: packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl:14]: This makes me sad, but I couldn't think of solution that doesn't involve running `pod install` during template instantiation which seemed... not great. Even with this limitation I think this is a huge improvement since "edit one text file" is a much easier thing for people with little-to-no Xcode experience than "make a new target and configure it correctly, then set up the test source structure including correct imports and test flow". But If you have ideas on how to make this unnecessary that would be great.

File: packages/flutter_tools/templates/template_manifest.json
Comment (stuartmorgan) [inline packages/flutter_tools/templates/template_manifest.json:57]: Drive-by cleanup since it confused my while renaming the existing file; this file doesn't actually exist any more (I assume since the ObjC/Swift split).

**Review by stuartmorgan-g — COMMENTED** [review]
This doesn't include macOS yet since I wanted to get your thoughts on this whole approach before replicating it (especially since the manual pxproj surgery is tedious).

File: packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl
Comment (jmagman) [inline packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl:1]: Should https://github.com/flutter/flutter/issues/51636 be closed?

File: packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl
Comment (jmagman) [reply packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl:18]: I'm okay with the conditionalizing in the project file as long as it's actually formatting the final project file correctly (not adding extra whitespace).

But how do you feel about always adding the `RunnerTests` to all new projects (iOS/macOS and Objective-C/Swift so we don't fragment further, though separate PRs okay), and the differences being the content of the test?  For example, for a plugin example project use your `FlutterMethodCall(methodName: "getPlatformVersion"` placeholder, and for a non-plugin app project it does something simple like assert `-[FlutterViewController setFlutterViewDidRenderCallback:]` callback is called.

The downside of this is the name of the test can't be `{{pluginClass}}Tests.swift` it would have to be something more generic.  Or we could keep `{{pluginClass}}Tests.swift` but that would involve keeping the conditionality for the test file.

File: packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl
Comment (jmagman) [reply packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl:428]: I don't think we should override target build settings that are already set at the project-level, like warnings, device family targeted (iPhone and iPad in this case), unless it's actually needed.  
Suggest:
```diff
				BUNDLE_LOADER = "$(TEST_HOST)";
-				CLANG_ANALYZER_NUMBER_OBJECT_CONVERSION = YES_AGGRESSIVE;
-				CLANG_CXX_LANGUAGE_STANDARD = "gnu++20";
-				CLANG_ENABLE_OBJC_WEAK = YES;
-				CLANG_WARN_DOCUMENTATION_COMMENTS = YES;
-				CLANG_WARN_QUOTED_INCLUDE_IN_FRAMEWORK_HEADER = YES;
-				CLANG_WARN_UNGUARDED_AVAILABILITY = YES_AGGRESSIVE;
				CODE_SIGN_STYLE = Automatic;
				CURRENT_PROJECT_VERSION = 1;
-				GCC_C_LANGUAGE_STANDARD = gnu11;
				GENERATE_INFOPLIST_FILE = YES;
-				IPHONEOS_DEPLOYMENT_TARGET = 11.0;
				MARKETING_VERSION = 1.0;
-				MTL_ENABLE_DEBUG_INFO = INCLUDE_SOURCE;
-				MTL_FAST_MATH = YES;
				PRODUCT_BUNDLE_IDENTIFIER = {{iosIdentifier}}.RunnerTests;
				PRODUCT_NAME = "$(TARGET_NAME)";
-				SWIFT_ACTIVE_COMPILATION_CONDITIONS = DEBUG;
-				SWIFT_EMIT_LOC_STRINGS = NO;
-				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
-				SWIFT_VERSION = 5.0;
-				TARGETED_DEVICE_FAMILY = "1,2";
				TEST_HOST = "$(BUILT_PRODUCTS_DIR)/Runner.app/$(BUNDLE_EXECUTABLE_FOLDER_PATH)/Runner";
```

File: packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl
Comment (jmagman) [reply packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl:14]: This problem would go away if we always add the `RunnerTest` since it would always be in the Podfile.

But even if that isn't done, can't we conditionalize in the Podfile in the same way?

File: packages/flutter_tools/templates/template_manifest.json
Comment (jmagman) [reply packages/flutter_tools/templates/template_manifest.json:57]: I think it does?

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/app_shared/ios.tmpl/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/commands.shard/permeable/create_test.dart:2423]:
```
03:49 +167 ~1 -1: test/commands.shard/permeable/create_test.dart: plugin includes native unit tests [E]                                                                                                
  Expected: a file system entity that exists
    Actual: LocalFile:<LocalFile: '/b/s/w/ir/x/t/flutter_tools.CSBNSE/flutter_tools_create_test.OVGLNS/flutter_project/example/ios/RunnerTests/FlutterProjectPluginTest.swift'>
     Which: does not exist
```

**Review by jmagman — COMMENTED** [review]
Suggest also adding a `runXcodeTests` call for iOS and macOS plugin example projects around here:  
https://github.com/flutter/flutter/blob/fc3571eff465da4e0131ef159ad10a3567d7b42f/dev/devicelab/lib/tasks/plugin_tests.dart#L61-L63
See example https://github.com/flutter/flutter/blob/8e2333dcc77bfa2b37ff28a5d6c80465ef515dc3/dev/devicelab/bin/tasks/native_platform_view_ui_tests_ios.dart#L49-L53
Can use `testWithNewIOSSimulator` deviceId to run tests on the simulator, see
https://github.com/flutter/flutter/blob/7c8e171320505c72444ff49a019f9d79c5cf170c/dev/devicelab/bin/tasks/module_test_ios.dart#L474

File: packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl
Comment (stuartmorgan) [reply packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl:18]: > But how do you feel about always adding the `RunnerTests` to all new projects (iOS/macOS and Objective-C/Swift so we don't fragment further, though separate PRs okay), and the differences being the content of the test?

The problem with that was that I couldn't think of any useful test to put there for non-plugin apps, so it seemed like it would be a weird target that nobody would actually use. But if we are okay with an empty placeholder that we don't think pretty much anyone will ever use I'm fine with that approach.

File: packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl
Comment (stuartmorgan) [reply packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl:14]: I forgot that we were creating it from a different template stage 🤦🏻  I'll take a look at parameterization options at that stage.

File: packages/flutter_tools/templates/template_manifest.json
Comment (stuartmorgan) [reply packages/flutter_tools/templates/template_manifest.json:57]: Hm, maybe I confused myself when looking locally. Since we also have -swift and -ios versions, I suspect this is a mistake (and is being overwritten maybe?) I'll investigate more.

File: packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl
Comment (stuartmorgan) [reply packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl:18]: > iOS/macOS and Objective-C/Swift so we don't fragment further, though separate PRs okay)

I'm definitely planning on doing it for all platforms and languages, I was just slicing the PRs. I'm happy to add ObjC here as well though.

Comment (jmagman) [reply packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl:18]: > The problem with that was that I couldn't think of any useful test to put there for non-plugin apps, so it seemed like it would be a weird target that nobody would actually use. But if we are okay with an empty placeholder that we don't think pretty much anyone will ever use I'm fine with that approach.

Once https://github.com/flutter/flutter/issues/56591 is fixed (someday) it would be great for the template to automatically set up integration_tests so running native tests would run and report the dart integration tests.  That would set up the same `RunnerTests` target for non-plugin example apps, so I'm fine with putting a placeholder there for now.

Comment (stuartmorgan) [reply packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl:18]: I was just talking to cbracken and mentioned this question, and he was fine with having native tests configured for macOS too, so I'll strip out all the conditionals. That will make this much less painful to preserve the target the next time we need to make changes to the project file.

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart
Comment (stuartmorgan) [reply packages/flutter_tools/test/commands.shard/permeable/create_test.dart:2423]: Sorry, I had roughed out the test but not actually run it yet, then forgot it was in there when I posted the draft for feedback.

File: packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl
Comment (stuartmorgan) [reply packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl]: It's now unconditional, and incudes ObjC.

File: packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl
Comment (stuartmorgan) [reply packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner.xcodeproj/project.pbxproj.tmpl:393]: Done (except for SWIFT_VERSION, which isn't set any higher for some reason).

File: packages/flutter_tools/templates/template_manifest.json
Comment (stuartmorgan) [reply packages/flutter_tools/templates/template_manifest.json:57]: It did exist, but shouldn't since it will overwrite the specific versions. I've removed the file as well now.

File: packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl
Comment (stuartmorgan) [reply packages/flutter_tools/templates/plugin_example/ios-swift.tmpl/RunnerTests/pluginClassTests.swift.tmpl]: I think that's about the Dart widget test? I'm not sure if we want a different standard for the Dart tests since they are higher profile in most cases (e.g., VS Code's Flutter plugin will surface it in the test inspector and auto-run it when running all tests).

**Comment by stuartmorgan** [comment]
Updated with everything but the device lab tests; I'll update again with those when I have a chance.

**Review by jmagman — COMMENTED** [review]
What you have LGTM, will review again when the devicelab tests are added.

**Comment by stuartmorgan** [comment]
Looks like the devicelab tests are working, thanks for all the pointers!

**Review by jmagman — APPROVED** [review]
LGTM!

**Pull request merged**
